src/OrbitBase/include/OrbitBase/JoinFutures.h

```diff
@@ -6,13 +6,88 @@
 #define ORBIT_BASE_JOIN_FUTURES_H_
 
 #include "OrbitBase/Future.h"
+#include "OrbitBase/FutureHelpers.h"
+#include "OrbitBase/Promise.h"
 #include "absl/types/span.h"
 
+namespace orbit_base_internal {
+
+template <typename T>
+class SharedStateJoin {
+  orbit_base::Promise<std::vector<T>> promise;
+  absl::Mutex mutex;
+  int incompleted_futures;
+  std::vector<std::optional<T>> results;
+
+  void SetResultsInPromise() {
+    std::vector<T> output;
+    output.reserve(results.size());
+    std::transform(results.begin(), results.end(), std::back_inserter(output),
+                   [](std::optional<T>& element) {
+                     CHECK(element.has_value());
+                     return *std::move(element);
```

Comment:
this line made me read documentation on the difference of value() and operator*. I did not know there was a difference. Now I know.

Reply:
Glad to be at service. :slightly_smiling_face:

Reply:
And yes. In this case it was intentional to call the "overload" that doesn't check for existence, since there is a CHECK above.

Reply:
Also the fact that value() can throw an exception makes me think we should avoid it entirely in our code and always use operator*